[git-webkit] Add screen-reader friendly review wizard (Part 1)

JonWBedard opened the pull request:
#### 26ef083bdcb33ab3c78a317860a6fa1b3aa3a9a1
<pre>
[git-webkit] Add screen-reader friendly review wizard (Part 1)
<a href="https://bugs.webkit.org/show_bug.cgi?id=261242">https://bugs.webkit.org/show_bug.cgi?id=261242</a>
<a href="https://rdar.apple.com/115083100">rdar://115083100</a>

Reviewed by Elliott Williams.

Add a read-only text-based PR review tool to git-webkit.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Add an &quot;editor&quot; argument which allows a caller to define
an editor callback which is invoke when the mock commit message editor is
called during the testing context.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py: Add Review program.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py: Added.
(Review.parser): Define arguments for the sub-program.
(Review.editor): Use a repository object to determine the user&apos;s prefered editor, matching
the editor used to edit commit messages.
(Review.args_for_url): Extract PR number and repository object from a URL, allowing a
user to specify a PR URL instead of a number and remote.
(Review.invoke_wizard): Generate a local file representing a PR, which includes the PR
diff and comments made against that diff. Open that local file in the user&apos;s prefered editor.
(Review.main): Extract a remote pull-request from the provided arguments and invoke the local
wizard with the pull-request details.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/review_unittest.py: Added.
(TestReview):
(TestReview.editor_callback):
(TestReview.editor_callback.callback):
(TestReview.test_bitbucket):
(TestReview.test_bitbucket_diff):
(TestReview.test_github):
(TestReview.test_github_files):
(TestReview.test_invalid_pr_url):
(TestReview.test_pr_argument):
(TestReview.test_editor_no_repo):
(TestReview.test_editor_repo):
(TestReview.test_invoke_wizard):
(TestReview.test_help):
(TestReview.test_bitbucket_read):
(TestReview.test_github_read):
(TestReview.test_bitbucket_read_comments):
(TestReview.test_github_read_comments):

Canonical link: <a href="https://commits.webkit.org/279244@main">https://commits.webkit.org/279244@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7983bbf2f11e1379f884c4e72d0f93dd345d661

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5399 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/56191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/3635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55217 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39001 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3360 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/56191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/3635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55010 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/29955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45669 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/56191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/52756 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2989 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1794 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3144 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57784 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/28053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/52919 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45886 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7768 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/29027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py:
```diff
@@ -0,0 +1,235 @@
+            help='Specify remote to search for pull request from.',
+        )
+        parser.add_argument(
+            '--dry-run', '-d', '--no-dry-run',
```
nit: "-n" is typically the single-character dry run argument.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py:
```diff
@@ -0,0 +1,235 @@
+        if from_config:
+            return from_config.split(' ')
+
+        from whichcraft import which
```
Why not use `shutil.which`? Surely we don't need to support python < 3.3 here.

JonWBedard replied:
Not any more! When this PR was first written, we still needed Python 2 support here.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py:
```diff
@@ -0,0 +1,235 @@
+            return from_config.split(' ')
+
+        from whichcraft import which
+        return [which('vi')]
```
nit: On some linux distributions (iirc, anything Debian-based), `vi` will be the ancient POSIX vi program, not vim. So you might want to default to vim?

JonWBedard committed:
[git-webkit] Add screen-reader friendly review wizard (Part 1)
https://bugs.webkit.org/show_bug.cgi?id=261242
rdar://115083100

Reviewed by Elliott Williams.

Add a read-only text-based PR review tool to git-webkit.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Add an "editor" argument which allows a caller to define
an editor callback which is invoke when the mock commit message editor is
called during the testing context.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py: Add Review program.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/review.py: Added.
(Review.parser): Define arguments for the sub-program.
(Review.editor): Use a repository object to determine the user's prefered editor, matching
the editor used to edit commit messages.
(Review.args_for_url): Extract PR number and repository object from a URL, allowing a
user to specify a PR URL instead of a number and remote.
(Review.invoke_wizard): Generate a local file representing a PR, which includes the PR
diff and comments made against that diff. Open that local file in the user's prefered editor.
(Review.main): Extract a remote pull-request from the provided arguments and invoke the local
wizard with the pull-request details.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/review_unittest.py: Added.
(TestReview):
(TestReview.editor_callback):
(TestReview.editor_callback.callback):
(TestReview.test_bitbucket):
(TestReview.test_bitbucket_diff):
(TestReview.test_github):
(TestReview.test_github_files):
(TestReview.test_invalid_pr_url):
(TestReview.test_pr_argument):
(TestReview.test_editor_no_repo):
(TestReview.test_editor_repo):
(TestReview.test_invoke_wizard):
(TestReview.test_help):
(TestReview.test_bitbucket_read):
(TestReview.test_github_read):
(TestReview.test_bitbucket_read_comments):
(TestReview.test_github_read_comments):

Canonical link: https://commits.webkit.org/279244@main
